posts/2024-07-12_teal_app_development_pharmaverseadam/teal-app-development.qmd

```diff
@@ -0,0 +1,208 @@
+shinyApp(app$ui, app$server)
+```
+
+### **Let’s Take a Closer Look at Each Step:**
+
+-   **In step 1**, we import libraries to use functions and datasets exported by those.
+
+-   **In step 2**, we create a teal data object that can be used in a teal app. There are a couple of subtle details here: instead of the default teal_data() function, we use a special wrapper around it, designed specifically for clinical trial data - cdisc_data(). Its advantage is that it will automatically generate join_keys() for the datasets that we pass in. Please note, that the dataset names should be all caps. The resulting data object is an s4 class instance, and we can verify that the output is correct by checking data names and join_keys slots:
+
+``` r
+r$> data@datanames
+[1] "ADAE" "ADSL"
+
+r$> data@join_keys
+A join_keys object containing foreign keys between 2 datasets:
+ADSL: [STUDYID, USUBJID]
+  <-- ADAE: [STUDYID, USUBJID]
+ADAE: [STUDYID, USUBJID, ASTDTM, AETERM, AESEQ]
+  --> ADSL: [STUDYID, USUBJID]
+```
+
+-   **In step 3**, we initialize the teal app by passing it a data object and adding an example module—it’s not much, but it will let us verify that the app works. The output app object is just a list with UI and server functions.
+
+-   **Finally step 4** should be familiar to all Shiny users - we need to pass the UI and server function from the app object generated at step 3.
+
+When we run the app, this is what we will see in the web browser:
+
+![](images/Asset 1.gif)
+
+As you can see, with just a few lines of code, we were able to create a working app with some interesting capabilities. We have two datasets that we can switch between. For each dataset, we have keys defined (marked with a special key icon in the variable dropdown). When we filter one dataset, the other one gets filtered as well because they are connected with a key.
+
+This is truly impressive, but there is just one problem… If we click the “Show R code” button, we will notice that the data used in the app is not “reproducible.” This simply means that the app currently does not have information about where the data comes from, so it cannot instruct users on how to obtain the same data.
+
+![](images/Asset 2.gif)
+
+Let’s fix this. We will have to make the code slightly more verbose:
+
+``` r
+data <- within(teal_data(), {
+  ADAE <- pharmaverseadam::adae
+  ADSL <- pharmaverseadam::adsl
+})
+datanames(data) <- c("ADAE", "ADSL")
+join_keys(data) <- default_cdisc_join_keys[datanames(data)]
+```
+
+The recommended method is to generate the data using a within function. However, this method requires manually providing databases using a helper function. We also need to provide `join_keys` ourselves, but given that the data names are standard ADaM names, we can take advantage of a special `default_cdisc_join_keys` object.
+
+It is also worth noting that [`{teal}`](https://insightsengineering.github.io/teal/latest-tag/) has its implementation of `within` generic.
```

Comment:
@bms63 should we suggest to use the {link} package, or leave it up to the author?